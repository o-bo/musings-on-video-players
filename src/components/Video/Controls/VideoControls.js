import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import {
  secondsToMinutesInWords,
  positionInsideProgressInSeconds,
} from './utils';

import './VideoControls.css';

/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/**
 * VideosControls displays the play/pause button and the progress bar.
 * It delegates actions to the {@link VideoContainer}.
 *
 * @extends PureComponent - in order to update only when props change.
 */
class VideoControls extends PureComponent {
  static propTypes = {
    currentProgress: PropTypes.number.isRequired,
    totalProgress: PropTypes.number.isRequired,
    playPause: PropTypes.func.isRequired,
    onTimeChange: PropTypes.func.isRequired,
    isPlaying: PropTypes.bool.isRequired,
  };

  /**
   * Computes the new time where the video should be started according to where
   * the user clicked on the progress bar.
   * Then it calls the onTimeChange method from the props - {@link VideoContainer#onTimeChange}.
   *
   * @param {object} e - the click event on the progrss bar.
   */
  changeCurrentTime = (e) => {
    const progressBoundingRect = this.progressBar.getBoundingClientRect();
    const setBoudingRectForDuration = positionInsideProgressInSeconds(progressBoundingRect);
    const computeTimeForPosition = setBoudingRectForDuration(this.props.totalProgress);
    const startTime = computeTimeForPosition(e.pageX);

    this.props.onTimeChange(startTime);
  }

  /**
   * Return the correct svg for playPauseIcon according to current props.
   */
  playPauseIcon = () => (this.props.isPlaying
    ? 'M14.016 5.016h3.984v13.969h-3.984v-13.969zM6 18.984v-13.969h3.984v13.969h-3.984z'
    : 'M8.016 5.016l10.969 6.984-10.969 6.984v-13.969z');

  render() {
    return (
      <div className="video-controls">
        <svg
          fill="white"
          width={23}
          height={23}
          onClick={this.props.playPause}
          style={{ padding: 5 }}
        >
          <path d={this.playPauseIcon()} />
        </svg>
        <span className="video-time">
          {secondsToMinutesInWords(parseInt(this.props.currentProgress, 10))}
        </span>
        <progress
          ref={(bar) => { this.progressBar = bar; }}
          value={this.props.currentProgress}
          max={this.props.totalProgress}
          onClick={this.changeCurrentTime}
        />
        <span className="video-time">
          {secondsToMinutesInWords(parseInt(this.props.totalProgress, 10))}
        </span>
      </div>
    );
  }
}
/* eslint-enable jsx-a11y/no-noninteractive-element-interactions */

export default VideoControls;
