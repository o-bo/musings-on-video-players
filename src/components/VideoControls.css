
progress {
	width: 70%;
	-webkit-appearance: none;
	border: none;
}

progress::-webkit-progress-bar {
	background: black;
	border-radius: 50px;
	padding: 2px;
	box-shadow: 0 1px 0px 0 rgba(255, 255, 255, 0.2);
}

progress::-webkit-progress-value {
	border-radius: 50px;
	box-shadow: inset 0 1px 1px 0 rgba(255, 255, 255, 0.4);
	background: #67DAF9;
}

.video-time {
  color: white;
  padding: 5px;
}
