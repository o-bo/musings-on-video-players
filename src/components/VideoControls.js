import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { elapsedTimeInWords } from './utils';

/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/**
 * VideosControls displays the play/pause button and the progress bar.
 * It delegates actions to the {@link VideoContainer}.
 *
 * @extends PureComponent - in order to update only when props change.
 */
class VideoControls extends PureComponent {
  static propTypes = {
    currentProgress: PropTypes.number.isRequired,
    totalProgress: PropTypes.number.isRequired,
    playPause: PropTypes.func.isRequired,
    onTimeChange: PropTypes.func.isRequired,
  };

  /**
   * Computes the new time where the video should be started according to where
   * the user clicked on the progress bar.
   * Then it calls the onTimeChange method from the props - {@link VideoContainer#onTimeChange}.
   *
   * @param {object} e - the click event on the progrss bar.
   */
  changeCurrentTime = (e) => {
    const progressBoundingRect = this.progressBar.getBoundingClientRect();
    const {
      width,
      left,
    } = progressBoundingRect;
    const offset = e.pageX - left;
    const startTime = (offset * this.props.totalProgress) / width;

    this.props.onTimeChange(startTime);
  }

  render() {
    return (
      <div>
        <button onClick={this.props.playPause}>Play/Pause</button>
        <progress
          ref={(bar) => { this.progressBar = bar; }}
          value={this.props.currentProgress}
          max={this.props.totalProgress}
          onClick={this.changeCurrentTime}
        />
        <span>
          {elapsedTimeInWords(this.props.currentProgress, this.props.totalProgress)}
        </span>
      </div>
    );
  }
}
/* eslint-enable jsx-a11y/no-noninteractive-element-interactions */

export default VideoControls;
