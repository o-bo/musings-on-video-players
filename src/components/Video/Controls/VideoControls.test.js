import renderer from 'react-test-renderer';
import React from 'react';
import {
  mount,
  shallow,
} from 'enzyme';

import VideoControls from './VideoControls';

const controlsParams = {
  currentProgress: 18,
  totalProgress: 60,
  playPause: jest.fn(),
  onTimeChange: jest.fn(),
  isPlaying: true,
};

it('renders a play icon with isPlaying at false', () => {
  const wrapper = shallow(<VideoControls {...controlsParams} isPlaying={false} />);
  expect(wrapper.instance().playPauseIcon()).toEqual('M8.016 5.016l10.969 6.984-10.969 6.984v-13.969z');
});

it('renders a pause icon with isPlaying at true', () => {
  const wrapper = shallow(<VideoControls {...controlsParams} />);
  expect(wrapper.instance().playPauseIcon()).toEqual('M14.016 5.016h3.984v13.969h-3.984v-13.969zM6 18.984v-13.969h3.984v13.969h-3.984z');
});

test('VideoControlsTest renders correctly in pause', () => {
  const tree = renderer.create(
    <VideoControls {...controlsParams} isPlaying={false} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('VideoControlsTest renders correctly in play', () => {
  const tree = renderer.create(
    <VideoControls {...controlsParams} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls onTimeChange when changeCurrentTime is called', () => {
  const wrapper = mount(<VideoControls {...controlsParams} />);
  wrapper.instance().changeCurrentTime({ pageX: 200 });
  expect(wrapper.prop('onTimeChange')).toHaveBeenCalled();
});
