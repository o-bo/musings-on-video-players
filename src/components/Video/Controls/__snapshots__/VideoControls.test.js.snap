// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoControlsTest renders correctly in pause 1`] = `
<div
  className="video-controls"
>
  <svg
    fill="white"
    height={23}
    onClick={[Function]}
    style={
      Object {
        "padding": 5,
      }
    }
    width={23}
  >
    <path
      d="M8.016 5.016l10.969 6.984-10.969 6.984v-13.969z"
    />
  </svg>
  <span
    className="video-time"
  >
    00:18
  </span>
  <progress
    max={60}
    onClick={[Function]}
    value={18}
  />
  <span
    className="video-time"
  >
    01:00
  </span>
</div>
`;

exports[`VideoControlsTest renders correctly in play 1`] = `
<div
  className="video-controls"
>
  <svg
    fill="white"
    height={23}
    onClick={[Function]}
    style={
      Object {
        "padding": 5,
      }
    }
    width={23}
  >
    <path
      d="M14.016 5.016h3.984v13.969h-3.984v-13.969zM6 18.984v-13.969h3.984v13.969h-3.984z"
    />
  </svg>
  <span
    className="video-time"
  >
    00:18
  </span>
  <progress
    max={60}
    onClick={[Function]}
    value={18}
  />
  <span
    className="video-time"
  >
    01:00
  </span>
</div>
`;
